api/v1alpha1/onload_types.go

```diff
@@ -4,25 +4,118 @@
 package v1alpha1
 
 import (
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
+// Currently unimplemented
+type KernelBuild struct {
+}
+
+type KernelMapping struct {
+	// Regexp is a regular expression that is used to match against the kernel
+	// versions of the nodes in the cluster
+	Regexp string `json:"regexp"`
+
+	// KernelImage is the image that contains the out-of-tree kernel modules
+	// used by onload, including the sfc driver.
+	KernelImage string `json:"kernelImage"`
+
+	// +optional
+	// Build contains build options for building kernel level components.
+	// If non-nil the operator will attempt to build the required images using
+	// what is defined in the build section, and push the images to the
+	// specified locations.
+	Build *KernelBuild `json:"build"`
+}
+
+// Currently unimplemented
+type OnloadBuildSpec struct {
+}
+
 // OnloadSpec defines the desired state of Onload
 type OnloadSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of Onload. Edit onload_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// KernelMappings is a list of pairs of kernel versions and container
+	// images. This allows for flexibility when there are heterogenous kernel
+	// versions on the nodes in the cluster.
+	KernelMappings []KernelMapping `json:"kernelMappings"`
```

Comment:
KMM (inc. KMM Hub, and indeed K8s in general) always defines container config in a sub `container` struct to avoid conflating the CRI abstraction of Pod/DaemonSet config with container engine Container config. We could push this granularity to a later milestone.